File: scripts/pre_commit_secret_scanner.py
"""
Pre-Commit Secret Scanner & Zero-Leak Enforcement Gate.
Scans all staged files before every git commit to guarantee that no credentials,
PostgreSQL connection strings with passwords, Neon keys, API tokens, or secrets
can ever be committed to the repository.
"""

import sys
import os
import re
import subprocess
from typing import List, Tuple

# Ensure UTF-8 output across all operating systems
if hasattr(sys.stdout, "reconfigure"):
    try:
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    except Exception:
        pass

# Patterns identifying real leaked credentials & connection strings
SECRET_PATTERNS = [
    (r"postgres(?:ql)?:\/\/[a-zA-Z0-9_]+:[a-zA-Z0-9_.\-~%!$&'()*+,;=]{6,}@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}", "PostgreSQL URI with hardcoded password"),
    (r"\bnpg_[a-zA-Z0-9]{10,}\b", "Neon Database Token (npg_*)"),
    (r"\bdapi[a-zA-Z0-9]{20,}\b", "Databricks Personal Access Token (dapi*)"),
    (r"\bhf_[a-zA-Z0-9]{20,}\b", "Hugging Face Access Token (hf_*)"),
    (r"\bsk-[a-zA-Z0-9_\-]{32,}\b", "OpenAI / Anthropic Secret Key (sk-*)"),
    (r"\bghp_[a-zA-Z0-9]{20,}\b", "GitHub Personal Access Token (ghp_*)"),
    (r"\bAIza[a-zA-Z0-9_\-]{35}\b", "Google API / Cloud Token"),
    (r"\bdp\.pt\.[a-zA-Z0-9]{20,}\b", "Doppler Service Token"),
    (r"-----BEGIN (?:RSA|EC|OPENSSH|PRIVATE) KEY-----", "Private Cryptographic Key"),
    (r"mongodb(?:\+srv)?:\/\/[a-zA-Z0-9_]+:[^@]+@", "MongoDB Connection String with Password"),
    (r"mysql:\/\/[a-zA-Z0-9_]+:[^@]+@", "MySQL Connection String with Password")
]

# Excluded directories and local-only config files
IGNORED_DIRS = {
    ".git", "node_modules", ".venv", "venv", "target", "dist", "build",
    ".pytest_cache", "__pycache__", ".turbo", ".next"
}

EXCLUDED_FILES = [
    "scripts/pre_commit_secret_scanner.py",
    ".env.example",
    ".env",
    "tests/"
]


def get_staged_files() -> List[str]:
    """Returns list of staged files in git."""
    try:
        res = subprocess.run(["git", "diff", "--cached", "--name-only"], capture_output=True, text=True, check=True)
        files = [f.strip() for f in res.stdout.splitlines() if f.strip()]
        return files
    except Exception:
        return []


def scan_file(filepath: str) -> List[Tuple[int, str, str]]:
    """Scans a single file for secret pattern violations."""
    if not os.path.exists(filepath) or os.path.isdir(filepath):
        return []
    
    # Check if excluded
    norm_path = filepath.replace("\\", "/")
    for exc in EXCLUDED_FILES:
        if exc in norm_path or norm_path == exc or norm_path == f"./{exc}":
            return []

    violations = []
    try:
        with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
            for line_idx, line in enumerate(f, start=1):
                # Ignore comment placeholders with dummy words
                if "user:pass" in line or "user:password" in line or "dummy" in line or "placeholder" in line or "mock_" in line or "example.invalid" in line:
                    continue
                for pattern, name in SECRET_PATTERNS:
                    if re.search(pattern, line):
                        violations.append((line_idx, name, line.strip()[:80]))
    except Exception:
        pass
    return violations


def main():
    staged = get_staged_files()
    if not staged:
        # If no staged files specified, scan all git-tracked and modified workspace files
        files_to_scan = []
        for root, dirs, files in os.walk("."):
            dirs[:] = [d for d in dirs if d not in IGNORED_DIRS]
            for file in files:
                files_to_scan.append(os.path.join(root, file))
        staged = files_to_scan

    total_violations = 0
    print("=" * 70)
    print("[SECURITY GATE] RUNNING PRE-COMMIT ZERO-LEAK SECRET SCANNER")
    print(f"Scanning {len(staged)} files...")
    print("=" * 70)

    for fpath in staged:
        violations = scan_file(fpath)
        if violations:
            for line_no, _, _ in violations:
                print(f"[BLOCKED] Security violation detected in {fpath} (Line {line_no})")
                total_violations += 1

    if total_violations > 0:
        print("\n" + "!" * 70)
        print(f"[CRITICAL ERROR] COMMIT ABORTED: {total_violations} potential secrets found in staged files!")
        print("Please remove all hardcoded tokens, passwords, or URIs before committing.")
        print("!" * 70)
        sys.exit(1)
    else:
        print("[OK] Zero secrets detected across all scanned files. Code is safe to commit.")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: scripts/pre_commit_secret_scanner.ts
/**
 * Pre-Commit Secret Scanner & Zero-Leak Enforcement Gate (Bun Runtime).
 * Scans all staged files before every git commit to guarantee that no credentials,
 * PostgreSQL connection strings with passwords, Neon keys, API tokens, or secrets
 * can ever be committed to the repository.
 */

import { existsSync, statSync, readFileSync, readdirSync } from "node:fs";
import { join, resolve } from "node:path";

interface SecretPattern {
  regex: RegExp;
  name: string;
}

const SECRET_PATTERNS: SecretPattern[] = [
  {
    regex: /postgres(?:ql)?:\/\/[a-zA-Z0-9_]+:[a-zA-Z0-9_.\-~%!$&'()*+,;=]{6,}@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/,
    name: "PostgreSQL URI with hardcoded password",
  },
  { regex: /\bnpg_[a-zA-Z0-9]{10,}\b/, name: "Neon Database Token (npg_*)" },
  { regex: /\bdapi[a-zA-Z0-9]{20,}\b/, name: "Databricks Personal Access Token (dapi*)" },
  { regex: /\bhf_[a-zA-Z0-9]{20,}\b/, name: "Hugging Face Access Token (hf_*)" },
  { regex: /\bsk-[a-zA-Z0-9_\-]{32,}\b/, name: "OpenAI / Anthropic Secret Key (sk-*)" },
  { regex: /\bghp_[a-zA-Z0-9]{20,}\b/, name: "GitHub Personal Access Token (ghp_*)" },
  { regex: /\bAIza[a-zA-Z0-9_\-]{35}\b/, name: "Google API / Cloud Token" },
  { regex: /\bdp\.pt\.[a-zA-Z0-9]{20,}\b/, name: "Doppler Service Token" },
  { regex: /-----BEGIN (?:RSA|EC|OPENSSH|PRIVATE) KEY-----/, name: "Private Cryptographic Key" },
  { regex: /mongodb(?:\+srv)?:\/\/[a-zA-Z0-9_]+:[^@]+@/, name: "MongoDB Connection String with Password" },
  { regex: /mysql:\/\/[a-zA-Z0-9_]+:[^@]+@/, name: "MySQL Connection String with Password" },
];

const IGNORED_DIRS = new Set([
  ".git", "node_modules", ".venv", "venv", "target", "dist", "build",
  ".pytest_cache", "__pycache__", ".turbo", ".next"
]);

const EXCLUDED_FILES = [
  "scripts/pre_commit_secret_scanner.py",
  "scripts/pre_commit_secret_scanner.ts",
  ".env.example",
  ".env",
  "tests/"
];

function getStagedFiles(): string[] {
  try {
    const proc = Bun.spawnSync(["git", "diff", "--cached", "--name-only"], {
      stdout: "pipe",
      stderr: "pipe",
    });
    const stdout = proc.stdout.toString().trim();
    if (!stdout) return [];
    return stdout.split(/\r?\n/).map((s) => s.trim()).filter(Boolean);
  } catch {
    return [];
  }
}

function scanFile(filepath: string): Array<{ lineNo: number; name: string }> {
  if (!existsSync(filepath)) return [];
  try {
    if (statSync(filepath).isDirectory()) return [];
  } catch {
    return [];
  }

  const normPath = filepath.replace(/\\/g, "/");
  for (const exc of EXCLUDED_FILES) {
    if (normPath.includes(exc) || normPath === exc || normPath === `./${exc}`) {
      return [];
    }
  }

  const violations: Array<{ lineNo: number; name: string }> = [];
  try {
    const content = readFileSync(filepath, "utf8");
    const lines = content.split(/\r?\n/);

    for (let idx = 0; idx < lines.length; idx++) {
      const line = lines[idx];
      if (
        line.includes("user:pass") ||
        line.includes("user:password") ||
        line.includes("dummy") ||
        line.includes("placeholder") ||
        line.includes("mock_") ||
        line.includes("example.invalid")
      ) {
        continue;
      }
      for (const pat of SECRET_PATTERNS) {
        if (pat.regex.test(line)) {
          violations.push({ lineNo: idx + 1, name: pat.name });
        }
      }
    }
  } catch {
    // Ignore binary or unreadable files
  }

  return violations;
}

function walkDir(dir: string): string[] {
  const files: string[] = [];
  try {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!IGNORED_DIRS.has(entry.name)) {
          files.push(...walkDir(join(dir, entry.name)));
        }
      } else {
        files.push(join(dir, entry.name));
      }
    }
  } catch {}
  return files;
}

function main(): void {
  let filesToScan = getStagedFiles();
  if (filesToScan.length === 0) {
    filesToScan = walkDir(".");
  }

  let totalViolations = 0;
  console.log("=".repeat(70));
  console.log("[SECURITY GATE] RUNNING PRE-COMMIT ZERO-LEAK SECRET SCANNER (BUN RUNTIME)");
  console.log(`Scanning ${filesToScan.length} files...`);
  console.log("=".repeat(70));

  for (const fpath of filesToScan) {
    const violations = scanFile(fpath);
    if (violations.length > 0) {
      for (const v of violations) {
        console.error(`[BLOCKED] Security violation detected in ${fpath} (Line ${v.lineNo}): ${v.name}`);
        totalViolations++;
      }
    }
  }

  if (totalViolations > 0) {
    console.error("\n" + "!".repeat(70));
    console.error(`[CRITICAL ERROR] COMMIT ABORTED: ${totalViolations} potential secrets found in scanned files!`);
    console.error("Please remove all hardcoded tokens, passwords, or URIs before committing.");
    console.error("!".repeat(70));
    process.exit(1);
  } else {
    console.log("[OK] Zero secrets detected across all scanned files. Code is safe to commit.");
    process.exit(0);
  }
}

main();
